sql/sql_class.h

```diff
@@ -1614,23 +1615,21 @@ class Bloom_filter {
      @return if true, the key might be a member of the set. If false, the key
      is definitely not a member of the set.
   */
-  bool test_and_set(MEM_ROOT *mem_root, ulong key)
+  bool test_and_set(ulong key)
   {
-    if (!bit_set)
+    Bit_set_ptr bit_set_copy= bit_set;
+    if (!bit_set_copy)
     {
-      void *bit_set_place= alloc_root(mem_root, sizeof(Bit_set));
-      // FIXME: memory allocation failure is eaten silently. Nonexact stats are
-      // the least of the concerns then.
-      if (bit_set_place == NULL)
-        return false;
-      bit_set= new (bit_set_place) Bit_set();
+      // TODO: possible inaccurate statistics because of races?
+      bit_set_copy= Bit_set_ptr(new Bit_set());
```

Comment:
I believe there is also boost::make_shared
```suggestion
      bit_set_copy= boost::make_shared<Bit_set>();
```